File: src/connectors/ip_api.js
import { BaseConnector } from './base.js';
import http from 'http';

export class IpApiConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'IP-API';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];
            const raw = await this._fetch(`http://ip-api.com/json/${encodeURIComponent(inputValue)}?fields=status,message,country,countryCode,region,regionName,city,zip,lat,lon,timezone,isp,org,as,asname,reverse,mobile,proxy,hosting,query`);
            let data;
            try { data = JSON.parse(raw); } catch { throw new Error('ip-api returned invalid JSON'); }

            if (data.status === 'success') {
                entities.push({
                    type: 'ip',
                    data: {
                        address: data.query,
                        country: data.countryCode,
                        country_name: data.country,
                        region: data.regionName,
                        city: data.city,
                        zip: data.zip,
                        lat: data.lat,
                        lon: data.lon,
                        timezone: data.timezone,
                        isp: data.isp,
                        org: data.org,
                        asn: data.as,
                        as_name: data.asname,
                        reverse_dns: data.reverse,
                        is_mobile: data.mobile,
                        is_proxy: data.proxy,
                        is_hosting: data.hosting,
                        source: 'ip-api'
                    },
                    confidence: 0.9
                });

                if (data.org && data.org !== data.isp) {
                    entities.push({
                        type: 'organization',
                        data: { name: data.org, source: 'ip_geolocation', asn: data.as },
                        confidence: 0.6
                    });
                }

                if (data.reverse) {
                    entities.push({
                        type: 'domain',
                        data: { name: data.reverse, source: 'reverse_dns' },
                        confidence: 0.7,
                        asKnown: { type: 'domain', value: data.reverse }
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _fetch(url) {
        return new Promise((resolve, reject) => {
            const req = http.get(url, { headers: { 'User-Agent': 'cyclops-osint' }, timeout: 10000 }, res => {
                if (res.statusCode === 429) { reject(new Error('ip-api rate limited')); res.resume(); return; }
                if (res.statusCode >= 400) { reject(new Error(`ip-api HTTP ${res.statusCode}`)); res.resume(); return; }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => resolve(data));
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('ip-api timeout')); });
        });
    }

    async healthCheck() { return { ok: true, version: 'native' }; }
}

File: cyclops/connectors/ip_api.py
import os
import aiohttp
from cyclops.connectors.base import BaseConnector


class IpApiConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "IP-API"
        self.api_key = os.environ.get("IP_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            fields = "status,message,country,countryCode,region,regionName,city,zip,lat,lon,timezone,isp,org,as,asname,reverse,mobile,proxy,hosting,query"
            if self.api_key:
                url = f"https://pro.ip-api.com/json/{input_value}?key={self.api_key}&fields={fields}"
            else:
                url = f"http://ip-api.com/json/{input_value}?fields={fields}"
            async with aiohttp.ClientSession() as session:
                async with session.get(url, timeout=aiohttp.ClientTimeout(total=10)) as resp:
                    data = await resp.json()
            entities = []
            if data.get("status") == "success":
                entities.append({"type": "ip", "data": {
                    "address": input_value, "country": data.get("country"), "country_code": data.get("countryCode"),
                    "region": data.get("regionName"), "city": data.get("city"), "lat": data.get("lat"), "lon": data.get("lon"),
                    "isp": data.get("isp"), "org": data.get("org"), "as": data.get("as"), "asname": data.get("asname"),
                    "reverse": data.get("reverse"), "is_proxy": data.get("proxy"), "is_mobile": data.get("mobile"),
                    "is_hosting": data.get("hosting"), "source": "ip_api",
                }, "confidence": 0.9})
                if data.get("org"):
                    entities.append({"type": "organization", "data": {"name": data["org"], "source": "ip_api"}, "confidence": 0.6})
                if data.get("reverse"):
                    entities.append({"type": "domain", "data": {"name": data["reverse"], "source": "ip_api_reverse"}, "confidence": 0.7,
                                     "as_known": {"type": "domain", "value": data["reverse"]}})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "hasApiKey": bool(self.api_key)}
